### ISSUE TITLE
RedisSlaveContext bug

### ISSUE THREAD

**Issue by charliboy** [riptideio/pymodbus#399]
### Versions

* Python:3.6.4
* OS:win 10
* Pymodbus:2.10
* Modbus Hardware (if used): 

### Pymodbus Specific
* Server: tcp/rtu/ascii - sync/async
* Client: tcp/rtu/ascii - sync/async

### Description

When using RedisSlaveContext, communication is not possible, and communication address validation is wrong 

### Code and Logs

```python
 # code and logs here.
#!/usr/bin/env python
"""
Pymodbus Synchronous Server Example
--------------------------------------------------------------------------

The synchronous server is implemented in pure python without any third
party libraries (unless you need to use the serial protocols which require
pyserial). This is helpful in constrained or old environments where using
twisted is just not feasible. What follows is an example of its use:
"""
# --------------------------------------------------------------------------- #
# configure the service logging
# --------------------------------------------------------------------------- #
#server code
import logging

from pymodbus.datastore import ModbusServerContext
from pymodbus.datastore.database import RedisSlaveContext
from pymodbus.device import ModbusDeviceIdentification
# --------------------------------------------------------------------------- #
# import the various server implementations
# --------------------------------------------------------------------------- #
from pymodbus.server.sync import StartTcpServer

FORMAT = ('%(asctime)-15s %(threadName)-15s'
          ' %(levelname)-8s %(module)-15s:%(lineno)-8s %(message)s')
logging.basicConfig(format=FORMAT)
log = logging.getLogger()
log.setLevel(logging.DEBUG)


def run_server():

    # ----------------------------------------------------------------------- #
    store = RedisSlaveContext( )
    context = ModbusServerContext(slaves=store, single=True)

    # ----------------------------------------------------------------------- #
    # initialize the server information
    # ----------------------------------------------------------------------- #
    # If you don't set this or any fields, they are defaulted to empty strings.
    # ----------------------------------------------------------------------- #
    identity = ModbusDeviceIdentification()
    identity.VendorName = 'Pymodbus'
    identity.ProductCode = 'PM'
    identity.VendorUrl = 'http://github.com/riptideio/pymodbus/'
    identity.ProductName = 'Pymodbus Server'
    identity.ModelName = 'Pymodbus Server'
    identity.MajorMinorRevision = '1.5'

    # ----------------------------------------------------------------------- #
    # run the server you want
    # ----------------------------------------------------------------------- #
    # Tcp:
    StartTcpServer(context, identity=identity, address=("localhost", 5020))

if __name__ == "__main__":
    run_server()

#client code
#!/usr/bin/env python
"""
Pymodbus Synchronous Client Examples
--------------------------------------------------------------------------

The following is an example of how to use the synchronous modbus client
implementation from pymodbus.

It should be noted that the client can also be used with
the guard construct that is available in python 2.5 and up::

    with ModbusClient('127.0.0.1') as client:
        result = client.read_coils(1,10)
        print result
"""
# --------------------------------------------------------------------------- #
# import the various server implementations
# --------------------------------------------------------------------------- #
from pymodbus.client.sync import ModbusTcpClient as ModbusClient
# from pymodbus.client.sync import ModbusUdpClient as ModbusClient
#from pymodbus.client.sync import ModbusSerialClient as ModbusClient

# --------------------------------------------------------------------------- #
# configure the client logging
# --------------------------------------------------------------------------- #
import logging
FORMAT = ('%(asctime)-15s %(threadName)-15s '
          '%(levelname)-8s %(module)-15s:%(lineno)-8s %(message)s')
logging.basicConfig(format=FORMAT)
log = logging.getLogger()
log.setLevel(logging.DEBUG)

UNIT = 0x1


def run_sync_client():

    # ------------------------------------------------------------------------#
    client = ModbusClient('localhost', port=5020)

    client.connect()

    log.debug("Reading Coils")
    rr = client.read_coils(1, 1, unit=UNIT)
    log.debug(rr)

    log.debug("Write to a Coil and read back")
    rq = client.write_coil(0, True, unit=UNIT)
    rr = client.read_coils(0, 1, unit=UNIT)
    assert(not rq.isError())     # test that we are not an error
    assert(rr.bits[0] == True)          # test the expected value

    log.debug("Write to multiple coils and read back- test 1")
    rq = client.write_coils(1, [True]*8, unit=UNIT)
    assert(not rq.isError())     # test that we are not an error
    rr = client.read_coils(1, 21, unit=UNIT)
    assert(not rr.isError())     # test that we are not an error
    resp = [True]*21

    # If the returned output quantity is not a multiple of eight,
    # the remaining bits in the final data byte will be padded with zeros
    # (toward the high order end of the byte).

    resp.extend([False]*3)
    assert(rr.bits == resp)         # test the expected value

    log.debug("Write to multiple coils and read back - test 2")
    rq = client.write_coils(1, [False]*8, unit=UNIT)
    rr = client.read_coils(1, 8, unit=UNIT)
    assert(not rq.isError())     # test that we are not an error
    assert(rr.bits == [False]*8)         # test the expected value

    log.debug("Read discrete inputs")
    rr = client.read_discrete_inputs(0, 8, unit=UNIT)
    assert(not rq.isError())     # test that we are not an error

    log.debug("Write to a holding register and read back")
    rq = client.write_register(1, 10, unit=UNIT)
    rr = client.read_holding_registers(1, 1, unit=UNIT)
    assert(not rq.isError())     # test that we are not an error
    assert(rr.registers[0] == 10)       # test the expected value

    log.debug("Write to multiple holding registers and read back")
    rq = client.write_registers(1, [10]*8, unit=UNIT)
    rr = client.read_holding_registers(1, 8, unit=UNIT)
    assert(not rq.isError())     # test that we are not an error
    assert(rr.registers == [10]*8)      # test the expected value

    log.debug("Read input registers")
    rr = client.read_input_registers(1, 8, unit=UNIT)
    assert(not rq.isError())     # test that we are not an error

    arguments = {
        'read_address':    1,
        'read_count':      8,
        'write_address':   1,
        'write_registers': [20]*8,
    }
    log.debug("Read write registeres simulataneously")
    rq = client.readwrite_registers(unit=UNIT, **arguments)
    rr = client.read_holding_registers(1, 8, unit=UNIT)
    assert(not rq.isError())     # test that we are not an error
    assert(rq.registers == [20]*8)      # test the expected value
    assert(rr.registers == [20]*8)      # test the expected value

    # ----------------------------------------------------------------------- #
    # close the client
    # ----------------------------------------------------------------------- #
    client.close()


if __name__ == "__main__":
    run_sync_client()

#server log:
2019-03-23 22:37:25,058 MainThread      DEBUG    sync           :317      Started thread to serve client at ('127.0.0.1', 12640)
2019-03-23 22:37:25,061 Thread-1        DEBUG    sync           :45       Client Connected [127.0.0.1:12640]
2019-03-23 22:37:25,062 Thread-1        DEBUG    sync           :171      Handling data: 0x0 0x1 0x0 0x0 0x0 0x6 0x1 0x1 0x0 0x1 0x0 0x1
2019-03-23 22:37:25,062 Thread-1        DEBUG    socket_framer  :147      Processing: 0x0 0x1 0x0 0x0 0x0 0x6 0x1 0x1 0x0 0x1 0x0 0x1
2019-03-23 22:37:25,062 Thread-1        DEBUG    factory        :130      Factory Request[1]
2019-03-23 22:37:25,062 Thread-1        DEBUG    redis_datastore:54       validate[1] 2:1
2019-03-23 22:37:26,094 Thread-1        ERROR    pdu            :107      Exception Response(129, 1, IllegalAddress)
2019-03-23 22:37:26,094 Thread-1        DEBUG    sync           :204      send: b'000100000003018102'
2019-03-23 22:37:26,094 Thread-1        DEBUG    sync           :171      Handling data: 0x0 0x2 0x0 0x0 0x0 0x6 0x1 0x5 0x0 0x0 0xff 0x0
2019-03-23 22:37:26,094 Thread-1        DEBUG    socket_framer  :147      Processing: 0x0 0x2 0x0 0x0 0x0 0x6 0x1 0x5 0x0 0x0 0xff 0x0
2019-03-23 22:37:26,094 Thread-1        DEBUG    factory        :130      Factory Request[5]
2019-03-23 22:37:26,094 Thread-1        DEBUG    redis_datastore:54       validate[5] 1:1
2019-03-23 22:37:26,094 Thread-1        ERROR    pdu            :107      Exception Response(133, 5, IllegalAddress)
2019-03-23 22:37:26,094 Thread-1        DEBUG    sync           :204      send: b'000200000003018502'
2019-03-23 22:37:26,094 Thread-1        DEBUG    sync           :171      Handling data: 0x0 0x3 0x0 0x0 0x0 0x6 0x1 0x1 0x0 0x0 0x0 0x1
2019-03-23 22:37:26,094 Thread-1        DEBUG    socket_framer  :147      Processing: 0x0 0x3 0x0 0x0 0x0 0x6 0x1 0x1 0x0 0x0 0x0 0x1
2019-03-23 22:37:26,094 Thread-1        DEBUG    factory        :130      Factory Request[1]
2019-03-23 22:37:26,094 Thread-1        DEBUG    redis_datastore:54       validate[1] 1:1
2019-03-23 22:37:26,094 Thread-1        ERROR    pdu            :107      Exception Response(129, 1, IllegalAddress)
2019-03-23 22:37:26,094 Thread-1        DEBUG    sync           :204      send: b'000300000003018102'
2019-03-23 22:37:26,109 Thread-1        DEBUG    sync           :171      Handling data: 
2019-03-23 22:37:26,125 Thread-1        DEBUG    socket_framer  :147      Processing: 
2019-03-23 22:37:26,125 Thread-1        DEBUG    sync           :53       Client Disconnected [127.0.0.1:12640]


#client log
2019-03-23 22:37:25,057 MainThread      DEBUG    testclient     :43       Reading Coils
2019-03-23 22:37:25,060 MainThread      DEBUG    transaction    :111      Current transaction state - IDLE
2019-03-23 22:37:25,060 MainThread      DEBUG    transaction    :115      Running transaction 1
2019-03-23 22:37:25,060 MainThread      DEBUG    transaction    :206      SEND: 0x0 0x1 0x0 0x0 0x0 0x6 0x1 0x1 0x0 0x1 0x0 0x1
2019-03-23 22:37:25,061 MainThread      DEBUG    sync           :75       New Transaction state 'SENDING'
2019-03-23 22:37:25,061 MainThread      DEBUG    transaction    :209      Changing transaction state from 'SENDING' to 'WAITING FOR REPLY'
2019-03-23 22:37:26,094 MainThread      DEBUG    transaction    :284      Changing transaction state from 'WAITING FOR REPLY' to 'PROCESSING REPLY'
2019-03-23 22:37:26,094 MainThread      DEBUG    transaction    :214      RECV: 0x0 0x1 0x0 0x0 0x0 0x3 0x1 0x81 0x2
2019-03-23 22:37:26,094 MainThread      DEBUG    socket_framer  :147      Processing: 0x0 0x1 0x0 0x0 0x0 0x3 0x1 0x81 0x2
2019-03-23 22:37:26,094 MainThread      DEBUG    factory        :246      Factory Response[129]
2019-03-23 22:37:26,094 MainThread      DEBUG    transaction    :363      Adding transaction 1
2019-03-23 22:37:26,094 MainThread      DEBUG    transaction    :374      Getting transaction 1
2019-03-23 22:37:26,094 MainThread      DEBUG    transaction    :180      Changing transaction state from 'PROCESSING REPLY' to 'TRANSACTION_COMPLETE'
2019-03-23 22:37:26,094 MainThread      DEBUG    testclient     :45       Exception Response(129, 1, IllegalAddress)
2019-03-23 22:37:26,094 MainThread      DEBUG    testclient     :47       Write to a Coil and read back
2019-03-23 22:37:26,094 MainThread      DEBUG    transaction    :111      Current transaction state - TRANSACTION_COMPLETE
2019-03-23 22:37:26,094 MainThread      DEBUG    transaction    :115      Running transaction 2
2019-03-23 22:37:26,094 MainThread      DEBUG    transaction    :206      SEND: 0x0 0x2 0x0 0x0 0x0 0x6 0x1 0x5 0x0 0x0 0xff 0x0
2019-03-23 22:37:26,094 MainThread      DEBUG    sync           :75       New Transaction state 'SENDING'
2019-03-23 22:37:26,094 MainThread      DEBUG    transaction    :209      Changing transaction state from 'SENDING' to 'WAITING FOR REPLY'
2019-03-23 22:37:26,094 MainThread      DEBUG    transaction    :284      Changing transaction state from 'WAITING FOR REPLY' to 'PROCESSING REPLY'
2019-03-23 22:37:26,094 MainThread      DEBUG    transaction    :214      RECV: 0x0 0x2 0x0 0x0 0x0 0x3 0x1 0x85 0x2
2019-03-23 22:37:26,094 MainThread      DEBUG    socket_framer  :147      Processing: 0x0 0x2 0x0 0x0 0x0 0x3 0x1 0x85 0x2
2019-03-23 22:37:26,094 MainThread      DEBUG    factory        :246      Factory Response[133]
2019-03-23 22:37:26,094 MainThread      DEBUG    transaction    :363      Adding transaction 2
2019-03-23 22:37:26,094 MainThread      DEBUG    transaction    :374      Getting transaction 2
2019-03-23 22:37:26,094 MainThread      DEBUG    transaction    :180      Changing transaction state from 'PROCESSING REPLY' to 'TRANSACTION_COMPLETE'
2019-03-23 22:37:26,094 MainThread      DEBUG    transaction    :111      Current transaction state - TRANSACTION_COMPLETE
2019-03-23 22:37:26,094 MainThread      DEBUG    transaction    :115      Running transaction 3
2019-03-23 22:37:26,094 MainThread      DEBUG    transaction    :206      SEND: 0x0 0x3 0x0 0x0 0x0 0x6 0x1 0x1 0x0 0x0 0x0 0x1
2019-03-23 22:37:26,094 MainThread      DEBUG    sync           :75       New Transaction state 'SENDING'
2019-03-23 22:37:26,094 MainThread      DEBUG    transaction    :209      Changing transaction state from 'SENDING' to 'WAITING FOR REPLY'
2019-03-23 22:37:26,094 MainThread      DEBUG    transaction    :284      Changing transaction state from 'WAITING FOR REPLY' to 'PROCESSING REPLY'
2019-03-23 22:37:26,094 MainThread      DEBUG    transaction    :214      RECV: 0x0 0x3 0x0 0x0 0x0 0x3 0x1 0x81 0x2
2019-03-23 22:37:26,094 MainThread      DEBUG    socket_framer  :147      Processing: 0x0 0x3 0x0 0x0 0x0 0x3 0x1 0x81 0x2
2019-03-23 22:37:26,094 MainThread      DEBUG    factory        :246      Factory Response[129]
2019-03-23 22:37:26,094 MainThread      DEBUG    transaction    :363      Adding transaction 3
2019-03-23 22:37:26,094 MainThread      DEBUG    transaction    :374      Getting transaction 3
2019-03-23 22:37:26,094 MainThread      DEBUG    transaction    :180      Changing transaction state from 'PROCESSING REPLY' to 'TRANSACTION_COMPLETE'
Traceback (most recent call last):
  File "C:/Users/LiPing/PycharmProjects/untitled/testclient.py", line 113, in <module>
    run_sync_client()
  File "C:/Users/LiPing/PycharmProjects/untitled/testclient.py", line 50, in run_sync_client
    assert(not rq.isError())     # test that we are not an error
AssertionError


```

**Comment by dhoomakethu** [comment]
Can you give me your redis schema ? And are there any entries in your redis db when you start polling ?

**Comment by charliboy** [comment]
> Can you give me your redis schema ? And are there any entries in your redis db when you start polling ?

Redis is the default configuration, not new schema. I understand that is the automatic generation when data is writing,But when the data is written from the client, the server will check the address first, and the server can not pass the address check.

**Comment by dhoomakethu** [comment]
Yes that is because the `RedisSlaveContext` does not initialise the db with default values and the db is essentially empty. If you can give a try by creating dummy entries with redis-cli `set pymodbus:h:<offset> <value>` and see if the writes works as expected. If you are using python3 , the read might not work after this and might end up with struct error. Having said that redis context purely experimental and is not complete. Feel free to play-around and raise a PR if you wish.

**Comment by charliboy** [comment]
@dhoomakethu ,thanks
I've corrected some bugs in the redis_datastore file and can now pass the above client code test in Python 3. But I haven't tested it in Python 2. 

#Redis database initialization code 
   host='localhost'
    port=6379
    r=redis.Redis(host=host, port=port)
    pre='pymodbus'
    num=100
    vals=[b'\x00\x00']*num
    c=('%s:c:%s' % (pre, v) for v in  range(num))
    c=dict(zip(c, vals))
    r.mset(c)
    d = ('%s:d:%s' % (pre, v) for v in range(num))
    d = dict(zip(d, vals))
    r.mset(d)
    vals=[0]*num
    h = ('%s:h:%s' % (pre, v) for v in range(num))
    h = dict(zip(h, vals))
    r.mset(h)
    i = ('%s:i:%s' % (pre, v) for v in range(num))
    i = dict(zip(i, vals))
    r.mset(i)
#redis_datastore.py
import redis
from pymodbus.interfaces import IModbusSlaveContext
from pymodbus.utilities import pack_bitstring, unpack_bitstring

# ---------------------------------------------------------------------------#
# Logging
# ---------------------------------------------------------------------------#
import logging;

_logger = logging.getLogger(__name__)


# ---------------------------------------------------------------------------#
# Context
# ---------------------------------------------------------------------------#
class RedisSlaveContext(IModbusSlaveContext):
    '''
    This is a modbus slave context using redis as a backing
    store.
    '''

    def __init__(self, **kwargs):
        ''' Initializes the datastores

        :param host: The host to connect to
        :param port: The port to connect to
        :param prefix: A prefix for the keys
        '''
        host = kwargs.get('host', 'localhost')
        port = kwargs.get('port', 6379)
        self.prefix = kwargs.get('prefix', 'pymodbus')
        self.client = kwargs.get('client', redis.Redis(host=host, port=port))
        self.__build_mapping()

    def __str__(self):
        ''' Returns a string representation of the context

        :returns: A string representation of the context
        '''
        return "Redis Slave Context %s" % self.client

    def reset(self):
        ''' Resets all the datastores to their default values '''
        self.client.flushall()

    def validate(self, fx, address, count=1):
        ''' Validates the request to make sure it is in range

        :param fx: The function we are working with
        :param address: The starting address
        :param count: The number of values to test
        :returns: True if the request in within range, False otherwise
        '''
        address = address + 1  # section 4.4 of specification
        _logger.debug("validate[%d] %d:%d" % (fx, address, count))
        return self.__val_callbacks[self.decode(fx)](address, count)

    def getValues(self, fx, address, count=1):
        ''' Validates the request to make sure it is in range

        :param fx: The function we are working with
        :param address: The starting address
        :param count: The number of values to retrieve
        :returns: The requested values from a:a+c
        '''
        address = address + 1  # section 4.4 of specification
        _logger.debug("getValues[%d] %d:%d" % (fx, address, count))
        return self.__get_callbacks[self.decode(fx)](address, count)

    def setValues(self, fx, address, values):
        ''' Sets the datastore with the supplied values

        :param fx: The function we are working with
        :param address: The starting address
        :param values: The new values to be set
        '''
        address = address + 1  # section 4.4 of specification
        _logger.debug("setValues[%d] %d:%d" % (fx, address, len(values)))
        self.__set_callbacks[self.decode(fx)](address, values)

    # --------------------------------------------------------------------------#
    # Redis Helper Methods
    # --------------------------------------------------------------------------#
    def __get_prefix(self, key):
        ''' This is a helper to abstract getting bit values

        :param key: The key prefix to use
        :returns: The key prefix to redis
        '''
        return "%s:%s" % (self.prefix, key)

    def __build_mapping(self):
        '''
        A quick helper method to build the function
        code mapper.
        '''
        self.__val_callbacks = {
            'd': lambda o, c: self.__val_bit('d', o, c),
            'c': lambda o, c: self.__val_bit('c', o, c),
            'h': lambda o, c: self.__val_reg('h', o, c),
            'i': lambda o, c: self.__val_reg('i', o, c),
        }
        self.__get_callbacks = {
            'd': lambda o, c: self.__get_bit('d', o, c),
            'c': lambda o, c: self.__get_bit('c', o, c),
            'h': lambda o, c: self.__get_reg('h', o, c),
            'i': lambda o, c: self.__get_reg('i', o, c),
        }
        self.__set_callbacks = {
            'd': lambda o, v: self.__set_bit('d', o, v),
            'c': lambda o, v: self.__set_bit('c', o, v),
            'h': lambda o, v: self.__set_reg('h', o, v),
            'i': lambda o, v: self.__set_reg('i', o, v),
        }

    # --------------------------------------------------------------------------#
    # Redis discrete implementation
    # --------------------------------------------------------------------------#
    __bit_size = 16
    __bit_default = b'\x00'

    def __get_bit_values(self, key, offset, count):
        ''' This is a helper to abstract getting bit values

        :param key: The key prefix to use
        :param offset: The address offset to start at
        :param count: The number of bits to read
        '''
        key = self.__get_prefix(key)
        s = divmod(offset, self.__bit_size)[0]
        e = divmod(offset + count, self.__bit_size)[0]

        request = ('%s:%s' % (key, v) for v in range(s, e + 1))
        response = self.client.mget(request)
        return response

    def __val_bit(self, key, offset, count):
        ''' Validates that the given range is currently set in redis.
        If any of the keys return None, then it is invalid.

        :param key: The key prefix to use
        :param offset: The address offset to start at
        :param count: The number of bits to read
        '''
        response = self.__get_bit_values(key, offset, count)
        return None not in response

    def __get_bit(self, key, offset, count):
        '''

        :param key: The key prefix to use
        :param offset: The address offset to start at
        :param count: The number of bits to read
        '''
        response = self.__get_bit_values(key, offset, count)
        print('pre', response)
        response = (r or self.__bit_default for r in response)
        result = b''.join(response)
        print('after', result)
        result = unpack_bitstring(result)
        return result[offset-1:offset + count-1]

    def __set_bit(self, key, offset, values):
        '''

        :param key: The key prefix to use
        :param offset: The address offset to start at
        :param values: The values to set
        '''
        count = len(values)
        s = divmod(offset, self.__bit_size)[0]
        e = divmod(offset + count, self.__bit_size)[0]
        current = self.__get_bit_values(key, offset, count)
        current_fmt = (r or self.__bit_default for r in current)
        if len(current) > 1:
            current = b''
            for el in current_fmt:
                current += el
        else:
            current = b''.join(current_fmt)
        current = unpack_bitstring(current)
        current[offset-1:offset + count-1] = values
        current = pack_bitstring(current)
        final = (current[s:s + self.__bit_size] for s in range(0, count, self.__bit_size))
        key = self.__get_prefix(key)
        request = ('%s:%s' % (key, v) for v in range(s, e + 1))
        request = dict(zip(request, final))
        self.client.mset(request)

    # --------------------------------------------------------------------------#
    # Redis register implementation
    # --------------------------------------------------------------------------#
    __reg_size = 16
    __reg_default = b'0'

    def __get_reg_values(self, key, offset, count):
        ''' This is a helper to abstract getting register values

        :param key: The key prefix to use
        :param offset: The address offset to start at
        :param count: The number of bits to read
        '''
        key = self.__get_prefix(key)
        # s = divmod(offset, self.__reg_size)[0]
        # e = divmod(offset+count, self.__reg_size)[0]

        # request  = ('%s:%s' % (key, v) for v in range(s, e + 1))
        request = ('%s:%s' % (key, v) for v in range(offset, offset+count))
        response = self.client.mget(request)
        return response

    def __val_reg(self, key, offset, count):
        ''' Validates that the given range is currently set in redis.
        If any of the keys return None, then it is invalid.

        :param key: The key prefix to use
        :param offset: The address offset to start at
        :param count: The number of bits to read
        '''
        response = self.__get_reg_values(key, offset, count)
        return None not in response

    def __get_reg(self, key, offset, count):
        '''

        :param key: The key prefix to use
        :param offset: The address offset to start at
        :param count: The number of bits to read
        '''
        response = self.__get_reg_values(key, offset, count)
        response = [int((r or self.__reg_default).decode()) for r in response]
        print('reg response',response)
        return response

    def __set_reg(self, key, offset, values):
        '''

        :param key: The key prefix to use
        :param offset: The address offset to start at
        :param values: The values to set
        '''
        count = len(values)
        # s = divmod(offset, self.__reg_size)
        # e = divmod(offset+count, self.__reg_size)

        # current = self.__get_reg_values(key, offset, count)

        key = self.__get_prefix(key)
        request = ('%s:%s' % (key, v) for v in range(offset, offset+count + 1))
        request = dict(zip(request, values))
        self.client.mset(request)

**Comment by dhoomakethu** [comment]
Thanks @charliboy I am working on some enhancements as well for redis store, I will update it here once thats ready.  Can you confirm if the reads are also working fine on Python3 ?

**Comment by charliboy** [comment]
@dhoomakethu ,Yes, at present, the above code works well on pyhon 3 and can pass the test  code on the client side.

**Comment by dhoomakethu** [comment]
Should be fixed with v2.2.0 and above
